File: sim/policies.py
"""The comparator policies, all driven through one identical loop (runner.py).

`gittins` is the engine under test, run on the real public path — encode
-> decide -> ledger learn/expire — never the layers in
isolation, so hashing, the epsilon-greedy rule, the tie split, and
forgetting are all in the loop together. The baselines bracket and
challenge it:

    oracle          — argmax of the true means: the regret upper bound
    uniform         — uniform random: the regret lower bound; together these
                      normalize regret to [0 (oracle) .. 1 (uniform)]
    greedy          — gittins with epsilon = 0 and first-max ties: same
                      model, same encoding, pure argmax of the estimates
    epsilon-greedy  — same model again, exploration as an epsilon coin
                      outside the engine: no ledger, first-max ties — what
                      the engine's plumbing is priced against

A policy is reset per run with `begin(seed)`; `choose` returns the chosen
candidate index plus its per-candidate reward estimates (None for the
model-free policies), which the runner uses for the prediction-RMSE metric;
`observe` hands back the reward. All policy randomness comes from sim.rand
streams keyed by (policy name, seed, t), so runs replay exactly.
"""

from gittins_reference.decide import decide, new_bandit
from gittins_reference.encoding import encode
from gittins_reference.exploration import DEFAULT_EPSILON
from gittins_reference.ledger import expire, learn
from gittins_reference.model import (
    DEFAULT_FORGETTING,
    factorize,
    new_model,
    predict_factored,
    update,
)
from sim.environments import Round
from sim.rand import randint, stream, uniform


class Policy:
    """One decision per round via `choose`, its reward via `observe`."""

    name: str

    def begin(self, seed: int) -> None:
        """Reset all run state; called once before each run."""

    def choose(self, rd: Round, seed: int) -> "tuple[int, list[float] | None]":
        """(chosen candidate index, per-candidate reward estimates or None)."""
        raise NotImplementedError

    def observe(self, reward: float) -> None:
        """The reward for the most recent `choose`; no-op for model-free
        policies."""


def argmax(values: "list[float] | tuple[float, ...]") -> int:
    """First maximum — the same tie-breaking rule as exploration.py."""
    best = 0
    for i in range(1, len(values)):
        if values[i] > values[best]:
            best = i
    return best


class OraclePolicy(Policy):
    """Upper bound; the only policy allowed to read Round.means."""

    name = "oracle"

    def choose(self, rd: Round, seed: int) -> "tuple[int, list[float] | None]":
        return argmax(rd.means), None


class UniformPolicy(Policy):
    """Lower bound: uniform over the candidates, no model."""

    name = "uniform"

    def choose(self, rd: Round, seed: int) -> "tuple[int, list[float] | None]":
        key = stream(self.name, seed, f"choose:{rd.t}")
        return randint(key, 0, len(rd.arm_ids)), None


class GittinsPolicy(Policy):
    """The engine, on the real public path: rewards are handed to the
    ledger's `learn` by decision id, and the `expire` sweep runs with every
    round's time, as a real event loop would. Rewards are immediate in the
    round runner, so nothing actually expires — the sweep is kept because
    it is part of the path a production loop drives."""

    def __init__(
        self,
        bits: int = 8,
        forgetting: float = DEFAULT_FORGETTING,
        horizon: float = 10.0,
        epsilon: float = DEFAULT_EPSILON,
    ):
        self.name = "gittins"
        self.bits = bits
        self.forgetting = forgetting
        self.horizon = horizon
        self.epsilon = epsilon

    def begin(self, seed: int) -> None:
        self.state = new_bandit(
            2**self.bits,
            horizon=self.horizon,
            epsilon=self.epsilon,
            forgetting=self.forgetting,
        )
        self.salt = f"{self.name}:{seed}"

    def choose(self, rd: Round, seed: int) -> "tuple[int, list[float] | None]":
        _, self.state = expire(self.state, rd.t)
        candidates = [
            encode(rd.context, rd.arm_ids[i], rd.actions[i], self.bits)
            for i in range(len(rd.arm_ids))
        ]
        record, self.state = decide(self.state, candidates, rd.t, self.salt)
        self._decision_id = record.decision_id
        self._t = rd.t
        # Metric-only read: the same estimates decide just scored with
        # (the model can't have changed between), recomputed because decide
        # deliberately logs only the chosen candidate.
        f = factorize(self.state.model)
        estimates = [predict_factored(f, x)[0] for x in candidates]
        return record.chosen, estimates

    def observe(self, reward: float) -> None:
        _, self.state = learn(self.state, self._decision_id, reward, self._t)


class ModelPolicy(Policy):
    """Shared machinery for the model-based baselines: the same
    per-coordinate forgetting ridge on the same hashed encoding as the
    engine, trained on every reward exactly as the engine trains — only the
    exploration rule differs, and there is no ledger anywhere."""

    def __init__(self, bits: int = 8, forgetting: float = DEFAULT_FORGETTING):
        self.bits = bits
        self.forgetting = forgetting

    def begin(self, seed: int) -> None:
        self.model = new_model(2**self.bits, self.forgetting)

    def estimates(self, rd: Round) -> "tuple[list, list[float]]":
        """(sparse candidates, their estimates) for one round."""
        candidates = [
            encode(rd.context, rd.arm_ids[i], rd.actions[i], self.bits)
            for i in range(len(rd.arm_ids))
        ]
        f = factorize(self.model)
        return candidates, [predict_factored(f, x)[0] for x in candidates]

    def observe(self, reward: float) -> None:
        self.model = update(self.model, self._chosen_x, reward)


class GreedyPolicy(ModelPolicy):
    """epsilon = 0, first-max ties: always the argmax estimate."""

    def __init__(self, bits: int = 8, forgetting: float = DEFAULT_FORGETTING):
        super().__init__(bits, forgetting)
        self.name = "greedy"

    def choose(self, rd: Round, seed: int) -> "tuple[int, list[float] | None]":
        candidates, estimates = self.estimates(rd)
        chosen = argmax(estimates)
        self._chosen_x = candidates[chosen]
        return chosen, estimates


class EpsilonGreedyPolicy(ModelPolicy):
    """Argmax estimate, except a uniform candidate with probability eps."""

    def __init__(self, eps: float, bits: int = 8, forgetting: float = DEFAULT_FORGETTING):
        super().__init__(bits, forgetting)
        if not (0.0 <= eps <= 1.0):
            raise ValueError("eps must be in [0, 1]")
        self.eps = eps
        self.name = f"epsilon-{eps}"

    def choose(self, rd: Round, seed: int) -> "tuple[int, list[float] | None]":
        candidates, estimates = self.estimates(rd)
        key = stream(self.name, seed, f"choose:{rd.t}")
        if uniform(key, 0) < self.eps:
            chosen = randint(key, 1, len(candidates))
        else:
            chosen = argmax(estimates)
        self._chosen_x = candidates[chosen]
        return chosen, estimates

File: core/src/bin/sim/policies.rs
//! The comparator policies — the port of `sim/policies.py`. `gittins` is
//! the engine on the real public path (encode -> decide -> ledger); the
//! baselines share its model and encoding and differ only in exploration.

use gittins_core::decide::{decide, new_bandit, BanditState};
use gittins_core::encoding::{encode, Features};
use gittins_core::exploration::DEFAULT_EPSILON;
use gittins_core::ledger::{expire, learn};
use gittins_core::model::{factorize, new_model, predict_factored, update, LinearModel, DEFAULT_FORGETTING};

use crate::environments::Round;
use crate::rand::{randint, stream, uniform};

pub trait Policy {
    fn name(&self) -> &str;
    /// Reset all run state; called once before each run.
    fn begin(&mut self, seed: u64);
    /// (chosen candidate index, per-candidate reward estimates or None).
    fn choose(&mut self, rd: &Round, seed: u64) -> (usize, Option<Vec<f64>>);
    /// The reward for the most recent `choose`; no-op for model-free policies.
    fn observe(&mut self, _reward: f64) {}
}

/// First maximum — the same tie-breaking rule as the reference's argmax.
fn argmax(values: &[f64]) -> usize {
    let mut best = 0;
    for i in 1..values.len() {
        if values[i] > values[best] {
            best = i;
        }
    }
    best
}

/// Upper bound; the only policy allowed to read Round.means.
pub struct OraclePolicy;

impl Policy for OraclePolicy {
    fn name(&self) -> &str {
        "oracle"
    }
    fn begin(&mut self, _seed: u64) {}
    fn choose(&mut self, rd: &Round, _seed: u64) -> (usize, Option<Vec<f64>>) {
        (argmax(&rd.means), None)
    }
}

/// Lower bound: uniform over the candidates, no model.
pub struct UniformPolicy;

impl Policy for UniformPolicy {
    fn name(&self) -> &str {
        "uniform"
    }
    fn begin(&mut self, _seed: u64) {}
    fn choose(&mut self, rd: &Round, seed: u64) -> (usize, Option<Vec<f64>>) {
        let key = stream("uniform", seed, &format!("choose:{:?}", rd.t));
        (randint(key, 0, rd.arm_ids.len()), None)
    }
}

/// Shared machinery for the model-based baselines: the same per-coordinate
/// forgetting ridge on the same hashed encoding as the engine.
struct ModelCore {
    bits: u32,
    model: LinearModel,
    chosen_x: Features,
}

impl ModelCore {
    fn new(bits: u32) -> ModelCore {
        ModelCore {
            bits,
            model: new_model(1 << bits, DEFAULT_FORGETTING, 1.0).unwrap(),
            chosen_x: Vec::new(),
        }
    }

    fn begin(&mut self) {
        self.model = new_model(1 << self.bits, DEFAULT_FORGETTING, 1.0).unwrap();
    }

    /// (sparse candidates, their estimates) for one round.
    fn estimates(&self, rd: &Round) -> (Vec<Features>, Vec<f64>) {
        let candidates: Vec<Features> = (0..rd.arm_ids.len())
            .map(|i| encode(&rd.context, &rd.arm_ids[i], &rd.actions[i], self.bits).unwrap())
            .collect();
        let mut f = factorize(&self.model);
        let estimates = candidates.iter().map(|x| predict_factored(&mut f, x).0).collect();
        (candidates, estimates)
    }

    fn observe(&mut self, reward: f64) {
        update(&mut self.model, &self.chosen_x, reward);
    }
}

/// epsilon = 0, first-max ties: always the argmax estimate.
pub struct GreedyPolicy {
    core: ModelCore,
}

impl GreedyPolicy {
    pub fn new(bits: u32) -> Self {
        GreedyPolicy { core: ModelCore::new(bits) }
    }
}

impl Policy for GreedyPolicy {
    fn name(&self) -> &str {
        "greedy"
    }
    fn begin(&mut self, _seed: u64) {
        self.core.begin();
    }
    fn choose(&mut self, rd: &Round, _seed: u64) -> (usize, Option<Vec<f64>>) {
        let (candidates, estimates) = self.core.estimates(rd);
        let chosen = argmax(&estimates);
        self.core.chosen_x = candidates.into_iter().nth(chosen).unwrap();
        (chosen, Some(estimates))
    }
    fn observe(&mut self, reward: f64) {
        self.core.observe(reward);
    }
}

/// Argmax estimate, except a uniform candidate with probability eps.
pub struct EpsilonGreedyPolicy {
    name: String,
    eps: f64,
    core: ModelCore,
}

impl EpsilonGreedyPolicy {
    pub fn new(eps: f64, bits: u32) -> Self {
        EpsilonGreedyPolicy {
            name: format!("epsilon-{eps}"),
            eps,
            core: ModelCore::new(bits),
        }
    }
}

impl Policy for EpsilonGreedyPolicy {
    fn name(&self) -> &str {
        &self.name
    }
    fn begin(&mut self, _seed: u64) {
        self.core.begin();
    }
    fn choose(&mut self, rd: &Round, seed: u64) -> (usize, Option<Vec<f64>>) {
        let (candidates, estimates) = self.core.estimates(rd);
        let key = stream(&self.name, seed, &format!("choose:{:?}", rd.t));
        let chosen = if uniform(key, 0, 0.0, 1.0) < self.eps {
            randint(key, 1, candidates.len())
        } else {
            argmax(&estimates)
        };
        self.core.chosen_x = candidates.into_iter().nth(chosen).unwrap();
        (chosen, Some(estimates))
    }
    fn observe(&mut self, reward: f64) {
        self.core.observe(reward);
    }
}

/// The engine, on the real public path: rewards go to the ledger's `learn`
/// by decision id, and the `expire` sweep runs with every round's time, as
/// a real event loop would.
pub struct GittinsPolicy {
    bits: u32,
    horizon: f64,
    state: BanditState,
    salt: String,
    decision_id: String,
    t: f64,
}

impl GittinsPolicy {
    pub fn new(bits: u32) -> Self {
        GittinsPolicy {
            bits,
            horizon: 10.0,
            state: new_bandit(1 << bits, 10.0, 0.0, DEFAULT_EPSILON, DEFAULT_FORGETTING).unwrap(),
            salt: String::new(),
            decision_id: String::new(),
            t: 0.0,
        }
    }
}

impl Policy for GittinsPolicy {
    fn name(&self) -> &str {
        "gittins"
    }
    fn begin(&mut self, seed: u64) {
        self.state = new_bandit(
            1 << self.bits,
            self.horizon,
            0.0,
            DEFAULT_EPSILON,
            DEFAULT_FORGETTING,
        )
        .unwrap();
        self.salt = format!("gittins:{seed}");
    }
    fn choose(&mut self, rd: &Round, _seed: u64) -> (usize, Option<Vec<f64>>) {
        expire(&mut self.state, rd.t);
        let candidates: Vec<Features> = (0..rd.arm_ids.len())
            .map(|i| encode(&rd.context, &rd.arm_ids[i], &rd.actions[i], self.bits).unwrap())
            .collect();
        let record = decide(&mut self.state, &candidates, rd.t, &self.salt, None, None).unwrap();
        self.decision_id = record.decision_id;
        self.t = rd.t;
        // Metric-only read: the same estimates decide just scored with,
        // recomputed because decide deliberately logs only the chosen
        // candidate.
        let mut f = factorize(&self.state.model);
        let estimates = candidates.iter().map(|x| predict_factored(&mut f, x).0).collect();
        (record.chosen, Some(estimates))
    }
    fn observe(&mut self, reward: f64) {
        learn(&mut self.state, &self.decision_id, reward, self.t);
    }
}
